### PR TITLE
[Feature branch] Introduce dataset source plugin system

### PR DESCRIPTION
**Pull request by dbczumar**
#xxx

## What changes are proposed in this pull request?

Introduce a plugin system for dataset sources in MLflow Tracking

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/data/dataset_source.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    def download(self) -> Any:
+        """
+        Downloads the files / objects referred to be this DatasetSource.
+
+        :return: The downloaded source, e.g. a local filesystem path, a Spark DataFrame, etc.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def _can_resolve(raw_source: Any) -> bool:
```
Comment (prithvikannan) [inline mlflow/data/dataset_source.py:35]: Just a thought -- would it make more sense to collapse `_can_resolve` with `_resolve` into something like `_resolve_if_possible`?

**Review by prithvikannan — COMMENTED** [review]
Really clear @dbczumar ! Think I found one small bug, and a few small questions about handling the case of multiple matching sources

File: mlflow/data/dataset_source_registry.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                matching_sources.append(source)
+
+        if len(matching_sources) > 1:
+            source_types_str = ", ".join([source._get_source_type() for source in matching_sources])
+            warnings.warn(
+                f"The specified dataset source can be interpreted in multiple ways:"
+                f" {source_types_str}. MLflow will assume that this is a"
+                f" {matching_sources[0]._get_source_type()} source.",
+                stacklevel=2,
+            )
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/data/dataset_source_registry.py:69]: Apologies for the lack of code comment / test here. The intention is to pick the last registered plugin, assuming that more general plugins are registered first and more specific plugins are registered last (e.g. third-party entrypoints). 

I'll add test coverage for this.

Added a comment here: https://github.com/mlflow/mlflow/pull/8046/files#diff-d9fe110424b274ff4515fb0149a551c1abdd185a475d973c855722d6fcb8fbaaR142-R149

File: mlflow/data/dataset_source_registry.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+        if len(matching_sources) >= 1:
+            return matching_sources[-1]._resolve(raw_source)
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/data/dataset_source_registry.py:72]: https://github.com/mlflow/mlflow/pull/8046/files#r1140463370

File: tests/data/test_dataset_source_registry.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    registry.resolve("candidate1:" + str(tmp_path))
+    registry.resolve("candidate1:" + str(tmp_path), candidate_sources=[CandidateDatasetSource1])
+    # CandidateDatasetSource1 is a subclass of TestDatasetSource and is therefore considered
+    # as a candidate for resolution
+    registry.resolve("candidate1:" + str(tmp_path), candidate_sources=[TestDatasetSource])
+    with pytest.raises(MlflowException, match="Could not find a source information resolver"):
+        # CandidateDatasetSource2 is not a superclass of CandidateDatasetSource1 or
+        # TestDatasetSource and cannot resolve raw sources with scheme "candidate1"
+        registry.resolve("candidate1:" + str(tmp_path), candidate_sources=[CandidateDatasetSource2])
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/data/test_dataset_source_registry.py:136]: Absolutely - adding a test case :)

File: mlflow/data/dataset_source_registry.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/data/dataset_source_registry.py:72]: Also fixed the warning message to use matching_sources[-1] and added a test case

File: tests/data/test_dataset_source_registry.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/data/test_dataset_source_registry.py:145]: Added test cases!

File: mlflow/data/dataset_source_registry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/data/dataset_source_registry.py]: This makes sense

**Review by prithvikannan — APPROVED** [review]
LGTM

**Pull request merged**